packages/fxa-content-server/app/scripts/models/account.js

```diff
@@ -372,6 +382,127 @@ const Account = Backbone.Model.extend(
+     *
+     * 1. Use kB if it is specified
+     * 2. Attempt to fetch keys if valid authentication token exists to do so
+     * 3. Use user's password to fetch keys
+     *
+     *
+     * @param {Object} [options={}]
+     *   @param {String} [options.kB] Optional The user's kB, ff specified takes priority.
+     *   @param {String} [options.password] Optional The user's password
+     * @returns {Promise}
+     */
+    async generateEcosystemAnonIdIfRequired(options = {}) {
+      try {
+        let ecosystemAnonId = this.get('ecosystemAnonId');
+        if (ecosystemAnonId) {
+          return;
+        }
+
+        // If no data pipeline keys are available, we can't generate the ecosystemAnonId, abort
+        if (this._ecosystemAnonIdPublicKeys.length <= 0) {
+          return;
+        }
+
+        // Pick a random public key
+        const randomEcosystemAnonIdPublicKey = this._ecosystemAnonIdPublicKeys[
+          Math.floor(Math.random() * this._ecosystemAnonIdPublicKeys.length)
+        ];
+
+        //
+        if (options.kB) {
+          ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+            options.kB,
+            randomEcosystemAnonIdPublicKey
+          );
+        } else if (this.canFetchKeys()) {
+          const keys = await this.accountKeys();
+          ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+            keys.kB,
+            randomEcosystemAnonIdPublicKey
+          );
+        } else if (options.password) {
+          // TODO: Using sessionReauth to get keys could potentially pose some issues if the user
+          // is attempting to come from an unverified session. FxA only allows fetching keys from
+          // sessions that are verified (ex. performed some email verification loop).
+          const res = await this._fxaClient.sessionReauth(
+            this.get('sessionToken'),
+            this.get('email'),
+            options.password,
+            {
+              wantsKeys: () => true, // Ensure keys are requested
+              has: () => false,
+              isSync: () => false,
+            },
+            {
+              keys: true,
+
+              // TODO: The verification reason needs to be updated to be specific for ecosystemAnonId
+              // There are checks in the auth-server to send emails depending on the `reason` value specified.
+              reason: VerificationReasons.RECOVERY_KEY,
+            }
+          );
+
+          if (res.keyFetchToken && res.unwrapBKey) {
+            const keys = await this._fxaClient.accountKeys(
+              res.keyFetchToken,
+              res.unwrapBKey
+            );
+            ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+              keys.kB,
+              randomEcosystemAnonIdPublicKey
+            );
+          }
+        }
+
+        // Update localstorage and report new ecosystemAnonId to auth-server
```

Comment:
Sorry to be a bit nitpicky here, but this is another comment that doesn't add to the meaning of the code below it. I think it'd be good to remove this one, too.